src/components/tl/sharp/tl_sharp_coll.c

```diff
@@ -0,0 +1,257 @@
+static ucc_status_t
+ucc_tl_sharp_mem_deregister(ucc_tl_sharp_context_t *ctx,
+                            ucc_tl_sharp_reg_t *reg)
+{
+    ucc_tl_sharp_rcache_region_t *region;
+
+    if (reg == &ucc_tl_sharp_reg_null) {
+        return UCC_OK;
+    }
+
+    if (ctx->rcache) {
+        region = ucc_container_of(reg, ucc_tl_sharp_rcache_region_t, reg);
+        ucc_rcache_region_put(ctx->rcache, &region->super);
+    } else {
+        sharp_coll_dereg_mr(ctx->sharp_context, reg->mr);
+        ucc_free(reg);
+    }
+
+    return UCC_OK;
+}
+
+ucc_status_t ucc_tl_sharp_collective_progress(ucc_coll_task_t *coll_task)
+{
+    ucc_tl_sharp_task_t *task  = ucc_derived_of(coll_task, ucc_tl_sharp_task_t);
+    int completed;
+
+    if (task->req_handle != NULL) {
+        completed = sharp_coll_req_test(task->req_handle);
+        if (completed) {
+            if (task->super.bargs.args.coll_type == UCC_COLL_TYPE_ALLREDUCE) {
+                ucc_tl_sharp_mem_deregister(TASK_CTX(task), task->allreduce.s_mem_h);
+                ucc_tl_sharp_mem_deregister(TASK_CTX(task), task->allreduce.r_mem_h);
+            }
+
+            sharp_coll_req_free(task->req_handle);
+            coll_task->super.status = UCC_OK;
+        }
+    }
+
+    return coll_task->super.status;
+}
+
+ucc_status_t ucc_tl_sharp_barrier_start(ucc_coll_task_t *coll_task)
+{
+    ucc_tl_sharp_task_t *task  = ucc_derived_of(coll_task, ucc_tl_sharp_task_t);
+    ucc_tl_sharp_team_t *team  = TASK_TEAM(task);
+    int ret;
+
+    task->super.super.status = UCC_INPROGRESS;
+
+    ret = sharp_coll_do_barrier_nb(team->sharp_comm, &task->req_handle);
+    if (ret != SHARP_COLL_SUCCESS) {
+        tl_error(UCC_TASK_LIB(task), "sharp_coll_do_barrier_nb failed:%s",
+                 sharp_coll_strerror(ret));
+        coll_task->super.status = UCC_ERR_NO_RESOURCE;
+        return ucc_task_complete(coll_task);
+    }
+
+    if (sharp_coll_req_test(task->req_handle) == 0) {
+        ucc_progress_enqueue(UCC_TL_CORE_CTX(team)->pq, &task->super);
+        return UCC_OK;
+    }
+
+    sharp_coll_req_free(task->req_handle);
+    coll_task->super.status = UCC_OK;
+    return ucc_task_complete(coll_task);
+}
+
+ucc_status_t ucc_tl_sharp_allreduce_start(ucc_coll_task_t *coll_task)
+{
+    ucc_tl_sharp_task_t *task  = ucc_derived_of(coll_task, ucc_tl_sharp_task_t);
+    ucc_tl_sharp_team_t *team  = TASK_TEAM(task);
+    ucc_coll_args_t     *args  = &TASK_ARGS(task);
+    size_t               count = args->src.info.count;
+    ucc_datatype_t       dt    = args->src.info.datatype;
+    struct sharp_coll_reduce_spec reduce_spec;
+    enum sharp_datatype sharp_type;
+    enum sharp_reduce_op op_type;
+    size_t data_size;
+    int ret;
+
+    task->super.super.status = UCC_INPROGRESS;
+
+    sharp_type = ucc_to_sharp_dtype[dt];
+    op_type    = ucc_to_sharp_reduce_op[args->reduce.predefined_op];
+    data_size  = ucc_dt_size(dt) * count;
+
+    ucc_tl_sharp_mem_register(TASK_CTX(task), args->src.info.buffer,
```

Comment:
maybe move mem_register to collective init? then we don't need to register/deregister buffer multiple times if we do multiple collective posts

Reply:
may be it is fine as we are using rcache here